### PR TITLE
use default payment term if BP does not have a payment term assigned

### PR DESCRIPTION
**Pull request by adi-stefan**

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java [hunk 1]
```diff
@@ -168,8 +168,11 @@ private I_C_Invoice saveInvoiceHeader(@NonNull final CreateManualInvoiceRequest
 		final int inputDataSourceRecordId = inputDataSourceDAO.retrieveInputDataSource(Env.getCtx(), API_DATA_SOURCE, /* throwEx */true, ITrx.TRXNAME_None)
 				.getAD_InputDataSource_ID();
 
-		invoiceRecord.setAD_Org_ID(createManualInvoiceRequest.getOrgId().getRepoId());
-		invoiceRecord.setC_BPartner_ID(createManualInvoiceRequest.getBillBPartnerId().getRepoId());
+		final OrgId orgId = createManualInvoiceRequest.getOrgId();
+		final BPartnerId billBPartnerId = createManualInvoiceRequest.getBillBPartnerId();
+		final SOTrx soTrx = createManualInvoiceRequest.getSoTrx();
+		invoiceRecord.setAD_Org_ID(orgId.getRepoId());
+		invoiceRecord.setC_BPartner_ID(billBPartnerId.getRepoId());
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java:175]: Is there any real change between these lines?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java [hunk 2]
```diff
@@ -179,10 +182,11 @@ private I_C_Invoice saveInvoiceHeader(@NonNull final CreateManualInvoiceRequest
 		invoiceRecord.setC_DocType_ID(createManualInvoiceRequest.getDocTypeId().getRepoId());
 		invoiceRecord.setC_DocTypeTarget_ID(createManualInvoiceRequest.getDocTypeId().getRepoId());
 		invoiceRecord.setPOReference(createManualInvoiceRequest.getPoReference());
-		invoiceRecord.setIsSOTrx(createManualInvoiceRequest.getSoTrx().toBoolean());
+		invoiceRecord.setIsSOTrx(soTrx.toBoolean());
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java:185]: is there any real change among these lines?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceService.java [hunk 3]
```diff
@@ -197,6 +201,11 @@ private CreateManualInvoiceRequest buildCreateManualInvoiceRequest(@NonNull fina
 		final BPartnerLocationAndCaptureId bPartnerLocationAndCaptureId = getBPartnerLocationAndCaptureId(requestHeader.getBillBPartnerLocationId());
 		final ZoneId zoneId = orgDAO.getTimeZone(requestHeader.getOrgId());
 		final PriceListId priceListId = getPriceListId(requestHeader, countryId, zoneId);
+		final Optional<PaymentTermId> paymentTermId = paymentTermRepository.retrievePaymentTermId(PaymentTermQuery.builder()
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceService.java:204]: why not inlining `paymentTermId`?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/request/CreateManualInvoiceRequest.java [hunk 4]
```diff
@@ -78,6 +79,8 @@ public class CreateManualInvoiceRequest
 	@NonNull
 	CurrencyId currencyId;
 
+	@NonNull PaymentTermId paymentTermId;
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/request/CreateManualInvoiceRequest.java:82]: i think this shall be nullable... see ur code changes from ManualInvoiceService

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java (hunk 1 above)
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java]: Nope, reverting

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java (hunk 2 above)
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceRepository.java]: Nope, reverted.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceService.java (hunk 3 above)
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/ManualInvoiceService.java]: Inlined.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java [hunk 5]
```diff
@@ -832,7 +833,10 @@ private IInvoiceHeader aggregate(final InvoiceHeaderAndLineAggregators headerAnd
 		}
 
 		invoiceHeader.setDocBaseType(docBaseType);
-		invoiceHeader.setPaymentTermId(getPaymentTermId(invoiceHeader).orElse(null));
+		final PaymentTermId paymentTermId = getPaymentTermId(invoiceHeader)
+				.orElse(paymentTermRepository.getDefaultPaymentTermId()
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/AggregationEngine.java]: switched to de.metas.util.Optionals#firstPresentOfSuppliers

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/request/CreateManualInvoiceRequest.java (hunk 4 above)
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/request/CreateManualInvoiceRequest.java:82]: C_Invoice.C_PaymentTerm_ID is mandatory. Changed logic in ManualInvoiceService.

**Review by teosarca — APPROVED** [review]

**Pull request merged**